language/move-prover/bytecode-to-boogie/src/lifetime_analysis.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    dataflow_analysis::{DataflowAnalysis, StateMap, TransferFunctions},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use bytecode_verifier::absint::{AbstractDomain, JoinResult};
+use stackless_bytecode_generator::stackless_bytecode::StacklessBytecode::{self, *};
+use std::collections::{BTreeMap, BTreeSet};
+use vm::file_format::{CodeOffset, LocalIndex, SignatureToken};
+
+// Represents a node in the borrow graph.
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum Node {
+    // Root means global storage or local storage. For example when there's BorrowLoc(dest, _) or
+    // BorrowGlobal(dest,_,_) then an edge will be added between Root -> Local(dest)
+    Root,
+
+    // Node representing a local. When BorrowField(dest, src), an edge between Local(src)
+    // and Local(dest) is added
+    Local(LocalIndex),
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct BorrowGraph {
+    graph: BTreeMap<Node, BTreeSet<Node>>,
+    reverse_graph: BTreeMap<Node, BTreeSet<Node>>,
+    moved_locals: BTreeSet<LocalIndex>,
+}
+
+impl BorrowGraph {
+    pub fn new() -> Self {
+        BorrowGraph {
+            graph: BTreeMap::new(),
+            reverse_graph: BTreeMap::new(),
+            moved_locals: BTreeSet::new(),
+        }
+    }
+
+    // Add an edge in the graph
+    pub fn add_edge(&mut self, from: Node, to: Node) {
+        // Fancy statement that modifies the value of key from if it exists, and inserts otherwise.
+        self.graph
+            .entry(from)
+            .and_modify(|e| {
+                e.insert(to);
+            })
+            .or_insert_with(|| [to].iter().cloned().collect());
```

Comment:
`|| vec![to.clone()]`?

Reply:
`vec![to.clone()]` gives you a vector but we want a `BTreeSet` here